File: src/main/java/mediathek/tool/ProgressMonitorInputStream.java
package mediathek.tool;

import org.jetbrains.annotations.NotNull;

import java.io.FilterInputStream;
import java.io.IOException;
import java.io.InputStream;

public class ProgressMonitorInputStream extends FilterInputStream {
    private final InputStreamProgressMonitor monitor;
    /**
     * The number of bytes that can be read from the InputStream.
     */
    private final long size;
    /**
     * The number of bytes that have been read from the InputStream.
     */
    private long bytesRead;

    /**
     * Creates a <code>FilterInputStream</code>
     * by assigning the  argument <code>in</code>
     * to the field <code>this.in</code> so as
     * to remember it for later use.
     *
     * @param in the underlying input stream, or <code>null</code> if
     *           this instance is to be created without an underlying stream.
     */
    public ProgressMonitorInputStream(InputStream in, long maxSize, InputStreamProgressMonitor mon) throws IOException {
        super(in);
        monitor = mon;
        this.size = maxSize;
        if (size == 0)
            throw new IOException("Size must be greater than zero!");
    }

    @Override
    public int read() throws IOException {
        final int read = super.read();
        if (read != -1) {
            bytesRead++;
            if (monitor != null)
                monitor.progress(bytesRead, size);
        }
        return read;
    }

    @Override
    public int read(byte @NotNull [] b) throws IOException {
        final int read = super.read(b);
        if (read != -1) {
            bytesRead += read;
            if (monitor != null)
                monitor.progress(bytesRead, size);
        }
        return read;
    }

    @Override
    public int read(byte @NotNull [] b, int off, int len) throws IOException {
        final int read = super.read(b, off, len);
        if (read != -1) {
            bytesRead += read;
            if (monitor != null)
                monitor.progress(bytesRead, size);
        }
        return read;
    }
}

File: src/main/kotlin/mediathek/tool/ProgressMonitorInputStream.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.tool

import java.io.FilterInputStream
import java.io.IOException
import java.io.InputStream

class ProgressMonitorInputStream(
    input: InputStream,
    private val size: Long,
    private val monitor: ((bytesRead: Long, size: Long) -> Unit)?,
) : FilterInputStream(input) {

    private var bytesRead = 0L

    init {
        if (size == 0L) {
            throw IOException("Size must be greater than zero!")
        }
    }

    @Throws(IOException::class)
    override fun read(): Int {
        val read = super.read()
        if (read != -1) {
            updateProgress(1)
        }
        return read
    }

    @Throws(IOException::class)
    override fun read(buffer: ByteArray): Int {
        val read = super.read(buffer)
        if (read != -1) {
            updateProgress(read.toLong())
        }
        return read
    }

    @Throws(IOException::class)
    override fun read(buffer: ByteArray, offset: Int, length: Int): Int {
        val read = super.read(buffer, offset, length)
        if (read != -1) {
            updateProgress(read.toLong())
        }
        return read
    }

    private fun updateProgress(readBytes: Long) {
        bytesRead += readBytes
        monitor?.invoke(bytesRead, size)
    }
}
